controllers/teleport_controller/teleport_controller.cpp

```diff
@@ -0,0 +1,147 @@
+    managers::RobotisOp2MotionManager mMotionManager = managers::RobotisOp2MotionManager(&supervisor);
+    mMotionManager.playPage(1);  // Look around
+
+    // Generate random seed
+    std::random_device rd;   // Will be used to obtain a seed for the random number engine
+    std::mt19937 gen(rd());  // Standard mersenne_twister_engine seeded with rd()
+    std::uniform_int_distribution<> xDistrib(0, 1080);
+    std::uniform_int_distribution<> yDistrib(0, 1600);
+
+    while (supervisor.step(timeStep) != -1) {
+
+        // Grab the current translation field of the robot to modify
+        webots::Field& target_translation_field = *target.getField("translation");
+        // Convert the field to a vector to output to console
+        const double* target_translation_vec = target_translation_field.getSFVec3f();
+
+        // Output current location
+        try {
+            std::ofstream log;
+            log.open("log.txt", std::fstream::app);
+            log << "Location: " << std::endl;
+            log << "X: " << target_translation_vec[0];
+            log << " Y: " << target_translation_vec[1];
+            log << " Z: " << target_translation_vec[2] << std::endl;
+            log.close();
+        }
+        catch (const std::string& e) {
+            std::cout << e << std::endl;
+        }
+        // Prepare new location
+
+        // 0, 0, 0 is centre of the playing field.
+        // X ranges from -5.4 - 5.4, the positive value on the left hand side when looking
+        // from red
+        // Y ranges from -8 - 8, the positive value on the red goal side
+        // Z should be set to 0.32 to be on the ground
+        // If the robot teleports into an existing object it may run into issues for that
+        // image only, after resetPhysics is should return to a regular state
+
+
+        double newPos[3];
+        newPos[0] = 5.4 - xDistrib(gen) / 100;
+        newPos[1] = 8 - yDistrib(gen) / 100;
+        newPos[2] = 0.32;
+
+        // Set new location
+        target_translation_field.setSFVec3f(newPos);
+
+        // Grab the current rotation field of the robot to modify
+        webots::Field& target_rotation_field = *target.getField("rotation");
+        // Convert the field to a vector to output to console
+        const double* target_rotation_vec = target_rotation_field.getSFRotation();
+
+        // Loop once for each rotation
+        for (long unsigned int i = 0; i < rotations.size(); i++) {
+            // Output current rotation
+            try {
+                std::ofstream log;
+                log.open("log.txt", std::fstream::app);
+                log << "Rotation: " << std::endl;
+                log << "X: " << target_rotation_vec[0];
+                log << " Y: " << target_rotation_vec[1];
+                log << " Z: " << target_rotation_vec[2];
+                log << " α: " << target_rotation_vec[3] << std::endl;
+                log.close();
```

Comment:
I am wondering if reopening this file will end up being a bottleneck for running this? probably not, since this generates, renders and saves a training image each iteration... but could be worth seeing if keeping the file open gives you a noticeable bump in performance (if we are finding this script is too slow later).

Reply:
I'll check it out